src/biogeophys/SnowCoverFractionMod.F90

```diff
@@ -0,0 +1,785 @@
+    class(snow_cover_fraction_ny07_type), intent(in) :: this
+    type(bounds_type), intent(in) :: bounds
+    integer, intent(in) :: num_c       ! number of columns in filter_c
+    integer, intent(in) :: filter_c(:) ! column filter to operate over
+
+    logical  , intent(in)    :: urbpoi( bounds%begc: )       ! true if the given column is urban
+    real(r8) , intent(in)    :: h2osno_total( bounds%begc: ) ! total snow water (mm H2O)
+    real(r8) , intent(in)    :: snowmelt( bounds%begc: )     ! total snow melt in the time step (mm H2O)
+    real(r8) , intent(in)    :: int_snow( bounds%begc: )     ! integrated snowfall (mm H2O)
+    real(r8) , intent(in)    :: newsnow( bounds%begc: )      ! total new snow in the time step (mm H2O)
+    real(r8) , intent(in)    :: bifall( bounds%begc: )       ! bulk density of newly fallen dry snow (kg/m3)
+
+    real(r8) , intent(inout) :: snow_depth( bounds%begc: )   ! snow height (m)
+    real(r8) , intent(inout) :: frac_sno( bounds%begc: )     ! fraction of ground covered by snow (0 to 1)
+    !
+    ! !LOCAL VARIABLES:
+    integer :: fc, c
+
+    character(len=*), parameter :: subname = 'UpdateSnowDepthAndFracNY07'
+    !-----------------------------------------------------------------------
+
+    SHR_ASSERT_FL((ubound(urbpoi, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(h2osno_total, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(snowmelt, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(int_snow, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(newsnow, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(bifall, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(snow_depth, 1) == bounds%endc), sourcefile, __LINE__)
+    SHR_ASSERT_FL((ubound(frac_sno, 1) == bounds%endc), sourcefile, __LINE__)
+
+    do fc = 1, num_c
+       c = filter_c(fc)
+
+       if (h2osno_total(c) > 0.0) then
+          snow_depth(c) = snow_depth(c) + newsnow(c) / bifall(c)
+          if(snow_depth(c) > 0.0_r8)  then
+             frac_sno(c) = tanh(snow_depth(c) / (2.5_r8 * this%zlnd * &
+                  (min(800._r8,(h2osno_total(c)+ newsnow(c))/snow_depth(c))/100._r8)**1._r8) )
+          end if
+          if(h2osno_total(c) < 1.0_r8)  then
+             frac_sno(c)=min(frac_sno(c),h2osno_total(c))
+          end if
+
+       else !h2osno_total == 0
+          ! initialize frac_sno and snow_depth when no snow present initially
+          if (newsnow(c) > 0._r8) then
+             snow_depth(c) = newsnow(c) / bifall(c)
+             if(snow_depth(c) > 0.0_r8)  then
+                frac_sno(c) = tanh(snow_depth(c) / (2.5_r8 * this%zlnd * &
+                     (min(800._r8,newsnow(c)/snow_depth(c))/100._r8)**1._r8) )
+             end if
+          else
+             snow_depth(c) = 0._r8
+             frac_sno(c) = 0._r8
+          end if
+       end if ! end of h2osno_total > 0
+    end do
```

Comment:
I wonder if this could be written in such a way to reduce redundancy?  Like this maybe:
  do fc = 1, num_c
       c = filter_c(fc)

! initialize frac_sno
    frac_sno(c) = 0._r8
! reset snow_depth if no h2osno    
    if (h2osno_total(c) == 0.0) snow_depth(c) = 0._r8
! update snow depth with new snowfall
   snow_depth(c) = snow_depth(c) + newsnow(c) / bifall(c)
! calculate frac_sno

          if(h2osno_total(c) < 1.0_r8)  then
             frac_sno(c)=min(frac_sno(c),h2osno_total(c))
	  else
	     if(snow_depth(c) > 0.0_r8)  then
             frac_sno(c) = tanh(snow_depth(c) / (2.5_r8 * this%zlnd * &
                  (min(800._r8,(h2osno_total(c)+ newsnow(c))/snow_depth(c))/100._r8)**1._r8) )
             end if
          end if
enddo

one thing I don't like is this though: if (h2osno_total(c) == 0.0) snow_depth(c) = 0._r8
it seems like a function doing the snow_depth reset should be called any time h2osno is 
updated, and then it wouldn't need to be done here.

Reply:
Good point. I'll take a stab at something like this. I'm not sure that the code you wrote is exactly right as is (particularly for the initial setting of frac_sno in the h2osno_total < 1 condition), but I see what you're getting at, and I can try for something similar.

The `h2osno_total` < 1 block makes this messier, and in looking more closely, I'm not sure it's really correct: although I don't understand what it's trying to do, it feels to me like this should be looking at (h2osno_total+newsnow), both in the condition and in the min, and that a similar min should be applied in the case where h2osno_total==0 but newsnow > 0. But I'm reluctant to make that change without really understanding it, so unless you give me guidance otherwise, I'll plan to keep the current code logic when doing the restructuring, even if this makes things less clean and consistent than they could be.

I agree that it feels like the snow_depth reset should happen somewhere else, but determining all the places that would need to happen would take some time; for now, I'll probably just add a note to look back at that.

Reply:
Fixed in 24ffe4e6. I have run a 3-year test (`SMS_Ly3.f10_f10_musgs.I2000Clm45BgcCrop.cheyenne_intel.clm-oldhyd_monthly`) and verified that it is bit-for-bit with master (despite the theoretical potential for this commit to change answers, as noted in the log message for the commit).

Reply:
@swensosc it might be good if you could look at the end result of 24ffe4e6 to make sure it (a) looks correct, and (b) accomplishes what you want.